google-cloud-bigtable/lib/google/cloud/bigtable/service.rb

```diff
@@ -83,16 +89,14 @@ def tables
         end
         attr_accessor :mocked_tables
 
-        def client
+        def client table_path, app_profile_id
           return mocked_client if mocked_client
-          @client ||= V2::Bigtable::Client.new do |config|
-            config.credentials = credentials if credentials
-            config.timeout = timeout if timeout
-            config.endpoint = host if host
-            config.lib_name = "gccl"
-            config.lib_version = Google::Cloud::Bigtable::VERSION
-            config.metadata = { "google-cloud-resource-prefix": "projects/#{@project_id}" }
+          table_key = "#{table_path}_#{app_profile_id}"
+          if @bigtable_clients.get(table_key).nil?
+            bigtable_client = create_bigtable_client table_path, app_profile_id
+            @bigtable_clients.put table_key, bigtable_client
           end
+          @bigtable_clients.get table_key
```

Comment:
Seems like we could encounter a race condition here if another thread pushes out the table_key after the check on line 95 but before line 99. You'll probably need to synchronize here.

Reply:
Updated to wrap it inside a mutex block.